File: buildSrc/src/main/java/com/tencent/strinker/RSymbols.java
package com.tencent.strinker;///*
// * Copyright (c) 2017 Yrom Wang
// *
// * Licensed under the Apache License, Version 2.0 (the "License");
// * you may not use this file except in compliance with the License.
// * You may obtain a copy of the License at
// *
// *    http://www.apache.org/licenses/LICENSE-2.0
// *
// * Unless required by applicable law or agreed to in writing, software
// * distributed under the License is distributed on an "AS IS" BASIS,
// * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// * See the License for the specific language governing permissions and
// * limitations under the License.
// */
//
//package com.tencent.strinker;
//
//import com.android.build.api.transform.DirectoryInput;
//import com.android.build.api.transform.TransformInput;
//import com.google.common.collect.HashBasedTable;
//import com.google.common.collect.Maps;
//import com.google.common.collect.Table;
//
//import org.objectweb.asm.ClassReader;
//import org.objectweb.asm.ClassVisitor;
//import org.objectweb.asm.FieldVisitor;
//import org.objectweb.asm.MethodVisitor;
//import org.objectweb.asm.Opcodes;
//
//import java.io.IOException;
//import java.io.UncheckedIOException;
//import java.nio.file.FileSystems;
//import java.nio.file.Files;
//import java.nio.file.Path;
//import java.nio.file.PathMatcher;
//import java.util.Arrays;
//import java.util.Collection;
//import java.util.Collections;
//import java.util.LinkedList;
//import java.util.List;
//import java.util.Map;
//import java.util.stream.Collectors;
//import java.util.stream.Stream;
//
//import static org.objectweb.asm.ClassReader.SKIP_DEBUG;
//import static org.objectweb.asm.ClassReader.SKIP_FRAMES;
//
///**
// * @author yrom
// */
//public class RSymbols {
//    /**
//     * default package!
//     */
//    static final String R_STYLEABLES_CLASS_NAME = "R$styleable";
//
//    private Map<String, Integer> symbols = Collections.emptyMap();
//    private Map<String, int[]> styleables = Maps.newHashMap();
//
//    // <owner class, field name, field value>
//    private Table<String, String, int[]> styleableTable = HashBasedTable.create();
//
//    private List<String> whiteList;
//    private List<String> whiteListRegex;
//
//    public RSymbols() {
//    }
//
//    public RSymbols(List<String> whiteList) {
//        this.whiteList = transformWhiteList(whiteList);
////        this.whiteListRegex = toRegex(whiteList);
//    }
//
//    /**
//     * @param name
//     * @param isClassName true means the first param is a class name like com.example.R$id.
//     *                    otherwise it is a field with class name like "com.example.R$id.some_id"
//     *                    fixme: bug when name is R$style or R$styleable
//     *                    fixme: using regex pattern
//     * @return
//     */
//    @Deprecated
//    public boolean isInWhiteList(String name, boolean isClassName) {
//        // using slash here for convenience to match the white list.
//        // com/example/R$id.some_id -> com/example/R/id/some_id
//        // com/example/R$id -> com/example/R/id
//        String key = name.replace('.', '/').replace('$', '/');
//        return whiteList != null
//                &&
//                (whiteList.stream()
//                        .filter(s -> s.endsWith("*"))
//                        .map(s -> s.substring(0, s.indexOf("*")))
//                        .anyMatch(s -> isClassName ? s.contains(key) : key.contains(s))
//                        ||
//                        whiteList.stream()
//                                .filter(s -> !s.endsWith("*"))
//                                .anyMatch(s -> isClassName ? s.contains(key) : key.equals(s))
//                );
//    }
//
////    private List<String> toRegex(List<String> origin) {
////        return origin.stream()
////                .peek(s -> {
////                    if (!s.matches(".*R\\.[a-zA-Z*]+\\..+")) {
////                        throw new RuntimeException(String.format("\"%s\" : Illegal format of white list! Please Check inlineR white list configuration!", s));
////                    }
////                })
////                .map(s -> {
////                    StringBuilder sb = new StringBuilder(s);
////                    if (!s.contains("$")) {
////                        sb.setCharAt(s.lastIndexOf(".", s.lastIndexOf(".") - 1), '$');
////                    }
////                    return sb.toString().replace(".", "/").replace("*", ".*").replace("$", "\\$");
////                })
////                .peek(s -> ShrinkerPlugin.logger.debug("[WhiteList]regex = {}", s))
////                .collect(Collectors.toList());
////    }
////
////    public boolean classNameMatchesWhiteList(String className) {
////        ShrinkerPlugin.logger.debug("[WhiteList]class name : {}", className);
////
////        return whiteListRegex != null
////                && whiteListRegex.stream().anyMatch(regex -> className.matches(regex.substring(0, regex.lastIndexOf("/"))));
//////                                                                                com/pkg/R$id
////    }
////
////    public boolean fieldNameMatchesWhiteList(String fieldName) {
////        String key = fieldName.replace('.', '/');
////        ShrinkerPlugin.logger.debug("[WhiteList]field name mapped for pattern: {} -> {}", fieldName, key);
////
////        return whiteListRegex != null
////                && whiteListRegex.stream().anyMatch(key::matches);
//////                                                  com/pkg/R$id/action
////    }
////
////    public boolean pathMatchesWhiteList2(String path) {
////        ShrinkerPlugin.logger.debug("[WhiteList]path : {}", path);
////
////        return whiteListRegex != null
////                && whiteListRegex.stream().anyMatch(regex -> path.matches(".*" + regex.substring(0, regex.lastIndexOf('/')) + "\\.class$"));
//////                                                                                .*com/pkg/R$id\.class$
////    }
//
//    @Deprecated
//    public boolean pathMatchesWhiteList(String path) {
//        String key = path.replace('.', '/').replace('$', '/');
//        return whiteList != null &&
//                whiteList.stream()
//                        .map(s -> s.endsWith("*")
//                                ? s.substring(0, s.indexOf("*"))
//                                : s.substring(0, s.lastIndexOf('/')))
//                        .anyMatch(key::contains);
//    }
//
//    public Integer get(String key) {
//        return symbols.get(key);
//    }
//
//    public boolean containsKey(String key) {
//        return symbols.containsKey(key);
//    }
//
//    public boolean isEmpty() {
//        return symbols.isEmpty() && styleables.isEmpty();
//    }
//
//    public Map<String, int[]> getStyleables() {
//        return Collections.unmodifiableMap(styleables);
//    }
//
//    public RSymbols from(Collection<TransformInput> inputs) {
//
//        final List<Path> paths = inputs.stream()
//                .map(TransformInput::getDirectoryInputs)
//                .flatMap(Collection::stream)
//                .map(this::toStream)
//                .reduce(Stream.empty(), Stream::concat)
//                .collect(Collectors.toList());
//        Stream<Path> stream;
//        if (paths.size() >= Runtime.getRuntime().availableProcessors() * 3) {
//            // use parallel here!
//            stream = paths.stream();
//            symbols = Maps.newConcurrentMap();
//        } else {
//            stream = paths.stream();
//            symbols = Maps.newHashMap();
//        }
//
//        final PathMatcher rClassMatcher = FileSystems.getDefault().getPathMatcher("glob:R$*.class");
//        stream.filter(path -> rClassMatcher.matches(path.getFileName()))
//                .forEach(this::drainSymbols);
//        return this;
//    }
//
//    private void drainSymbols(Path file) {
////        final String filename = file.getFileName().toString();
////        String typeName = filename.substring(0, filename.length() - ".class".length());
//        byte[] bytes;
//        try {
//            bytes = Files.readAllBytes(file);
//        } catch (IOException e) {
//            throw new UncheckedIOException(e);
//        }
//
//        ClassVisitor visitor = new ClassVisitor(Opcodes.ASM5) {
//
//            String className;
//
//            @Override
//            public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
//                super.visit(version, access, name, signature, superName, interfaces);
//                className = name;
//            }
//
//            @Override
//            public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
//                // read constant value
//                if (value instanceof Integer) {
//                    String key = className + '.' + name;
//                    Integer old = symbols.get(key);
//                    if (old != null && !old.equals(value)) {
//                        throw new IllegalStateException("Value of " + key + " mismatched! "
//                                + "Excepted 0x" + Integer.toHexString(old)
//                                + " but was 0x" + Integer.toHexString((Integer) value));
//                    } else {
////                        if (fieldNameMatchesWhiteList(key)) {
////                            return null;
////                        }
//                        symbols.put(key, (Integer) value);
//                    }
//                }
//                return null;
//            }
//
//            @Override
//            public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
//                if (access == Opcodes.ACC_STATIC && "<clinit>".equals(name)) {
//
//                    return new MethodVisitor(Opcodes.ASM5) {
//                        int[] current = null;
//                        boolean inError;
//                        LinkedList<Integer> intStack = new LinkedList<>();
//
//                        @Override
//                        public void visitIntInsn(int opcode, int operand) {
//                            if (opcode == Opcodes.NEWARRAY && operand == Opcodes.T_INT) {
//                                if (intStack.size() > 0) {
//                                    current = new int[intStack.pop()];
//                                } else {
//                                    inError = true;
//                                    System.err.println(String.format("%s : %s -> %s", className, "visitIntInsn", "NEWARRAY inError"));
//                                }
//                            } else if (opcode == Opcodes.BIPUSH || opcode == Opcodes.SIPUSH) {
//                                intStack.push(operand);
//                            }
//                        }
//
//                        @Override
//                        public void visitIincInsn(int var, int increment) {
//                            super.visitIincInsn(var, increment);
//                        }
//
//                        @Override
//                        public void visitVarInsn(int opcode, int var) {
//                            super.visitVarInsn(opcode, var);
//                        }
//
//                        @Override
//                        public void visitLdcInsn(Object cst) {
//                            if (cst instanceof Integer) {
//                                intStack.push((Integer) cst);
//                            }
//                        }
//
//                        @Override
//                        public void visitInsn(int opcode) {
//                            if (opcode >= Opcodes.ICONST_0 && opcode <= Opcodes.ICONST_5) {
//                                intStack.push(opcode - Opcodes.ICONST_0);
//                            } else if (opcode == Opcodes.IASTORE) {
//                                if (!inError) {
//                                    int value = intStack.pop();
//                                    int index = intStack.pop();
//                                    if (current != null) {
//                                        current[index] = value;
//                                    }
//                                }
//                            }
//                        }
//
//                        @Override
//                        public void visitFieldInsn(int opcode, String owner, String name, String desc) {
//                            if (opcode == Opcodes.PUTSTATIC) {
//                                if (!inError) {
//                                    int[] old = styleables.get(name);
//                                    if (old != null && old.length != current.length && !Arrays.equals(old, current)) {
//                                        throw new IllegalStateException("Value of styleable." + name + " mismatched! "
//                                                + "Excepted " + Arrays.toString(old)
//                                                + " but was " + Arrays.toString(current));
//                                    } else {
//                                        styleables.put(name, current);
//                                    }
//                                }
//                                inError = false;
//                                current = null;
//                                intStack.clear();
//                            }
//                        }
//                    };
//                }
//                return null;
//            }
//        };
//
//        new ClassReader(bytes).accept(visitor, SKIP_DEBUG | SKIP_FRAMES);
//    }
//
//    private List<String> transformWhiteList(List<String> origin) {
//        return origin.stream()
//                .map(s -> s.replace('.', '/'))
//                .collect(Collectors.toList());
//    }
//
////    private List<String> toRegex(List<String> origin) {
////        return origin.stream().map(s -> s.replace(".","\\.").replace("$","\\$"))
////                .collect(Collectors.toList());
////    }
//
//    private Stream<Path> toStream(DirectoryInput dir) {
//        try {
//            return Files.walk(dir.getFile().toPath()).filter(Files::isRegularFile);
//        } catch (IOException e) {
//            throw new UncheckedIOException(e);
//        }
//    }
//}

File: buildSrc/src/main/java/com/tencent/strinker/RSymbols.kt
package com.tencent.strinker

import com.android.build.api.transform.DirectoryInput
import com.android.build.api.transform.TransformInput
import org.objectweb.asm.*
import java.io.IOException
import java.io.UncheckedIOException
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream
import kotlin.streams.toList

/**
 *
 *
 */
class RSymbols() {

    companion object {
        /**
         * default package!
         */
        const val R_STYLEABLES_CLASS_NAME = "R\$styleable"


        /**
         * 从inputs中解析出所有的R$xxx类
         */
        fun collectAllRFiles(inputs: Collection<TransformInput>): RSymbols {
            val symbols = RSymbols()

            val paths = inputs.stream()
                    // 所有依赖的包都会在主工程下生成一个对应的R文件
                    .map {
                        it.directoryInputs
                    }
                    //
                    .flatMap<DirectoryInput> {
                        it.stream()
                    }
                    .map<Stream<Path>> {
                        toStream(it)
                    }
                    .reduce(Stream.empty()) { stream, stream1 ->
                        Stream.concat(stream, stream1)
                    }
                    .collect(Collectors.toList())


            val stream: Stream<Path>
            if (paths.size >= Runtime.getRuntime().availableProcessors() * 3) {
                // use parallel here!

                stream = paths.parallelStream()
                symbols.symbols = mutableMapOf()
            } else {
                stream = paths.stream()
                symbols.symbols = mutableMapOf()
            }

            // 找到所有R相关的类
            val rClassMatcher = FileSystems.getDefault().getPathMatcher("glob:R$*.class")
            stream.filter {
                rClassMatcher.matches(it.fileName)
            } .forEach {
                drainRSymbolsFromRFile(it, symbols)
            }

            return symbols
        }




        private fun drainRSymbolsFromRFile(file: Path, symbols: RSymbols) {

            val filename = file.fileName.toString()
            val typeName = filename.substring(0, filename.length - ".class".length)

            // 读取文件内容
            val bytes: ByteArray
            try {
                bytes = Files.readAllBytes(file)
            } catch (e: IOException) {
                throw UncheckedIOException(e)
            }

            val visitor = object : ClassVisitor(Opcodes.ASM5) {

                // 收集非R$.styleables的R$xxx类
                override fun visitField(access: Int, name: String?, desc: String?, signature: String?, value: Any?): FieldVisitor? {
                    // read constant value
                    if (value !is Int) return null // 如果不是int值，比表示

                    val key = "$typeName.$name"
                    val old = symbols.symbols[key]

                    if (old != null && old != value) {
                        throw IllegalStateException("Value of " + key + " mismatched! "
                                + "Excepted 0x" + Integer.toHexString(old)
                                + " but was 0x" + Integer.toHexString(value))
                    } else {
                        symbols.symbols[key] = value
                    }
                    return null
                }

                override fun visitMethod(access: Int, name: String?, desc: String?, signature: String?, exceptions: Array<String>?): MethodVisitor? {

                    return if (access == Opcodes.ACC_STATIC && "<clinit>" == name) {

                        object : MethodVisitor(Opcodes.ASM5) {

                            var current: IntArray? = null
                            var intStack = LinkedList<Int>()

                            override fun visitIntInsn(opcode: Int, operand: Int) {
                                if (opcode == Opcodes.NEWARRAY && operand == Opcodes.T_INT) {
                                    current = IntArray(intStack.pop())
                                } else if (opcode == Opcodes.BIPUSH || opcode == Opcodes.SIPUSH) {
                                    intStack.push(operand)
                                }
                            }

                            override fun visitLdcInsn(cst: Any) {
                                if (cst is Int) {
                                    intStack.push(cst)
                                }
                            }

                            override fun visitInsn(opcode: Int) {
                                if (opcode >= Opcodes.ICONST_0 && opcode <= Opcodes.ICONST_5) {
                                    intStack.push(opcode - Opcodes.ICONST_0)
                                } else if (opcode == Opcodes.IASTORE) {
                                    val value = intStack.pop()
                                    val index = intStack.pop()
                                    current?.set(index, value)
                                }
                            }

                            override fun visitFieldInsn(opcode: Int, owner: String, name: String, desc: String) {
                                if (opcode == Opcodes.PUTSTATIC) {
                                    val old = symbols.styleables[name]
                                    if (old != null && old.size != current!!.size && !Arrays.equals(old, current)) {
                                        throw IllegalStateException("Value of styleable." + name + " mismatched! "
                                                + " but was " + Arrays.toString(current))
                                    } else {
                                        if (current != null)
                                            symbols.styleables[name] = current!!
                                    }
                                    current = null
                                    intStack.clear()
                                }
                            }
                        }
                    } else null
                }
            }



            //
            ClassReader(bytes).accept(visitor, ClassReader.SKIP_DEBUG or ClassReader.SKIP_FRAMES)
        }


        private fun toStream(dir: DirectoryInput): Stream<Path> {
            try {
                return Files.walk(dir.file.toPath()).filter {path -> Files.isRegularFile(path) }
            } catch (e: IOException) {
                throw UncheckedIOException(e)
            }
        }
    }

    /**
     * 比如R.string.a=0xd7060001，保存在map里应为为["string.a"] = 0xd7060001
     */
    private var symbols: MutableMap<String, Int> = mutableMapOf()

    /**
     * 比如R.styleables.b = int[] {1, 2}，保存在map里应该为[styleables.b] = int[] {1, 2}
     */
    private val styleables = mutableMapOf<String, IntArray>()

    val isEmpty: Boolean
        get() = symbols.isEmpty() && styleables.isEmpty()

    operator fun get(key: String): Int? {
        return symbols[key]
    }

    fun containsKey(key: String): Boolean {
        return symbols.containsKey(key)
    }

    fun getStyleables(): Map<String, IntArray> {
        return Collections.unmodifiableMap(styleables)
    }
}
